src/ModelGradient.cpp

```diff
@@ -151,11 +151,84 @@ void LRSparseGradient::serialize(void* mem) const {
   }
 }
 
+/**
+ * Method will serialize like this [version(int), num_weights, data to send ps1,
+ * version(int), num_weights, data to send ps2, ... data to send ps_parts]
+ *
+ */
+
+std::vector<std::tuple<int, int>> LRSparseGradient::shard_serialize(
+    void* mem,
+    uint32_t parts) const {
+  std::vector<int> starts(parts, 0);
+  std::vector<std::tuple<int, int>> starts_out(parts);
+
+  // Perform count
+  for (const auto& w : weights) {
+    starts[w.first % parts]++;
```

Comment:
This is the line where we distribute the weights across servers, right? I think we should hash the weight index.

That way we avoid clusters of weights that can create imbalances in terms of the distribution, which creates bottlenecks on the servers. In my scalability experiments, I did see some imbalance in the network utilization that may stem from this.

Reply:
I believe this does hash the weight index. w.first is the index for weight w.second, right? I'll double check this

Reply:
Nevermind, I misread that. I'll hash these.

Reply:
Current work in progress...

Reply:
Fixed